README.md
git-deploy
==========

It is a tool to manage using git as a deployment management tool implemented in Python and utilizing the dulwich [1] project for native git functionality in Python.

[1] https://github.com/jelmer/dulwich


Usage
-----

Basic usage involves cloning the remote working repo to the deploy target and all client nodes.  When
a client is ready to deploy 'start' is invoked to obtain a lock on the remote and 'sync' is called to
push changes to the target.  On completion the lock is removed.

To start a new deployment, navigate to the working repo, issue a 'start' command:

    $ git deploy start [opts]

At this point you are free to make commits to the project and when ready for deployment issue 
a 'sync' command - or simply 'sync' if you're work is already complete but, be sure to rebase
your local clone:

    $ git deploy sync [opts]

The process may be aborted at any point with an 'abort' command:

    $ git deploy abort [opts]

You can rollback to a tag with the revert call:

    $ git deploy revert [-t <tag_name>] [opts]

If no tag is supplied the rollback uses the most recent tag.  The default is to only commit the rollback locally
however, by suppling the "-a" option for auto-sync the rollback automatically syncs also:

    $ git deploy revert [-t <tag_name>] [opts]

Deploy Hooks
------------

In the working path of your local clone deploy hooks may be added to '.git/deploy/hooks'.  You are
free to write your own hooks however, simple default hooks have been provided in 'git-deploy/default-hooks',
these can be copied to the hook folder in each client and target.


Setup
-----

The following example illustrates how git-deploy can be configured and used for a single client and target.  For this
example the client host and target host will be referred to as *client.realm.org* and *target.realm.org* respectively.
The home path on the client and target for our sample project is */home/me/project/* and */var/www/html/sample.com/*
respectively.  It is assumed that the git remote is configured for your project and for this example the remote alias is
*origin* and the remote branch is *master*.


**CLIENT SETUP**

The client is any working machine from which deployments may be initiated.  On *client.realm.org* clone git deploy, local
install, configure settings and initialize:

    $ git clone git@github.com:Git-Tools/git-deploy.git

    ...

    $ cd git-deploy

    $ sudo pip install -e .

    ...

Next configure the client instance with git config by assigning the following settings in *scripts/git-deploy.ini*:

    [deploy]

    target=target.realm.org

    path=/home/me/project/

    user=me

    hook-dir=.git/deploy/hooks/

    tag-prefix=sample.com

    remote=origin

    branch=master

    client-path=/home/me/project/

    key-path=/home/me/.ssh/id_rsa

    test-repo-path=/tmp/test_repo/

    [system]

    run_root=/usr/bin/

Once you have defined settings in *git-deploy.ini* call *init.py* to set the got config

    $ sudo ./scripts/init_gd


**TARGET SETUP**

On *target.realm.org* there is no need to clone and install git-deploy but here the deploy hooks will need to be
created.  There is a default hook in *git-deploy/default-hooks/default-target-pull.py* that should be copied to
*target.realm.org:/var/www/html/sample.com/.git/deploy/hooks/*.  This is a basic hook that will pull the changes
pushed from the client instance on sync.


**USING GIT DEPLOY**

First initialize a new repository on *client.realm.org*:

    client.realm.org:~ me$ mkdir me.com

    client.realm.org:~ me$ cd me.com

    client.realm.org:~ me$ git init

    client.realm.org:~ me$ git remote add origin git@github.com:wikimedia/me.com.git

    client.realm.org:~ me$ git push origin master

Next initialize the client repo by following the client setup above.  Subsequently, initialize the deploy target
on *target.realm.org* as indicated.

    target.realm.org:~ me$ cp ~/default-client-pull.py /var/www/html/sample.com/.git/deploy/hooks/

    target.realm.org:~ me$ chmod +x /var/www/html/sample.com/.git/deploy/hooks/default-client-pull.py


*Start* & *Sync*:

Ensure that the client is correctly synced to the remote by issuing a git pull or rebase.  Then you can issue a
a start command to write the lock file to the target to begin the deployment.

    client.realm.org:~ me$ cd me.com

    client.realm.org:me.com me$ touch new_file

    client.realm.org:me.com me$ git add new_file

    client.realm.org:me.com me$ git commit -m "add - my new file"

    client.realm.org:me.com me$ git pull --rebase origin master

At this point you are ready to enter the deployment process:

    client.realm.org:me.com me$ git deploy start

    <perform any testing or add any other commits as necessary you've locked the remote at this point>

    client.realm.org:me.com me$ git deploy sync

Once you sync a the specified hooks will be invoked from the client and the target and a tag is written to the
repository on the latest commit of the deploy. If the default push and pull hooks are used the client will simply
push its changes to the remote and the target will pull in the new changes.  Deployment tags have the form
*<repo>-sync-<date>-<time>*.


*Abort*

At times it is necessary to exit the deploy cycle prematurely.  For instance, consider the following:

    client.realm.org:me.com me$ git deploy start

    client.realm.org:me.com me$ git commit bad_change -m "add - some buggy code."

Suddenly, you realize that your change introduced a bug after entering the deloy process.  Rather than syncing the bad
code and then rolling back (next section) we can simply abort the deploy:

    client.realm.org:me.com me$ git deploy abort

    client.realm.org:me.com me$ git reset --soft HEAD^

    ... continue with your local changes ...

Now you have released deploy to other clients without infecting the code base with your buggy code.


*Rollback*

If you accidentally deploy some code that needs to be rolled back the *revert* command can be very helpful here.  You
can rollback to previous deploy states by utilizing deploy tags.  To view the old tags:

    client.realm.org:me.com me$ git tag

Now to rollback to a previous deploy call *git revert* with the appropriate tag:

    client.realm.org:me.com me$ git deploy start

    client.realm.org:me.com me$ git deploy revert <tag>


Examples
--------

**Reverting to a tag**

This example illustrates how to rollback a deploy to an earlier tag.

In the directory of the client repository check Git history:

    richrushbay-lm:test_sartoris rfaulk$ git log

    commit 779a07774dd5a7baf8dc86657cbfc491264ff970
    Author: rfaulk <rfaulk@yahoo-inc.com>
    Date:   Sun Dec 8 23:28:50 2013 -0800

        test

    commit 1ab80a78fc1e89ae6f8872282f96f5f42677b843
    Author: rfaulk <rfaulk@yahoo-inc.com>
    Date:   Sat Nov 16 18:31:26 2013 -0800

        test git-deploy 20131116_0.file.

    commit 12311863e111a266c4c1c513da45a39ed3e8cdd5
    Author: rfaulk <rfaulk@yahoo-inc.com>
    Date:   Sat Nov 2 19:42:58 2013 -0700

        Initial commit.

Next, take a look at the existing sync tags:

    richrushbay-lm:test_sartoris rfaulk$ git tag

    sartoris-sync-20131220-000105
    sartoris-sync-20131220-012354
    sartoris-sync-20131220-013503

Let's start the deploy:

    richrushbay-lm:test_sartoris rfaulk$ git deploy start

    Dec-20 01:53:00 DEBUG    git-deploy is ready to run
    Dec-20 01:53:00 INFO     git_deploy.git_deploy :: Config - {'deploy.test_repo': '/tmp/test/', 'deploy.key_path': '/Users/rfaulk/.ssh/id_rsa', 'target': 'stat1.wikimedia.org', 'top_dir': '/Users/rfaulk/Projects/test_sartoris', 'user.name': 'rfaulk', 'user.email': 'rfaulk@yahoo-inc.com', 'deploy.remote_url': 'git@github.com:rfaulkner/test_sartoris.git', 'hook_dir': '.git/deploy/hooks/', 'client_path': '/Users/rfaulk/Projects/test_sartoris/', 'sync_dir': '.git/deploy/hooks//sync', 'path': '/home/rfaulk/test_sartoris/', 'repo_name': 'sartoris', 'deploy_file': '/Users/rfaulk/Projects/test_sartoris/.git/.deploy', 'user': 'rfaulk'}
    Dec-20 01:53:00 INFO     git_deploy.git_deploy :: Checking for lock file at stat1.wikimedia.org.
    Dec-20 01:53:01 INFO     git_deploy.git_deploy :: No lock file exists.
    Dec-20 01:53:01 INFO     git_deploy.git_deploy :: SSH Lock create.

Next call the revert action - a tag can be explicitly supplied, but in this example it isn't and the rollback applies to the
previous sync tag:

    richrushbay-lm:test_sartoris rfaulk$ git deploy revert

    Dec-20 01:53:13 DEBUG    git-deploy is ready to run
    Dec-20 01:53:13 INFO     git_deploy.git_deploy :: Config - {'deploy.test_repo': '/tmp/test/', 'deploy.key_path': '/Users/rfaulk/.ssh/id_rsa', 'target': 'stat1.wikimedia.org', 'top_dir': '/Users/rfaulk/Projects/test_sartoris', 'user.name': 'rfaulk', 'user.email': 'rfaulk@yahoo-inc.com', 'deploy.remote_url': 'git@github.com:rfaulkner/test_sartoris.git', 'hook_dir': '.git/deploy/hooks/', 'client_path': '/Users/rfaulk/Projects/test_sartoris/', 'sync_dir': '.git/deploy/hooks//sync', 'path': '/home/rfaulk/test_sartoris/', 'repo_name': 'sartoris', 'deploy_file': '/Users/rfaulk/Projects/test_sartoris/.git/.deploy', 'user': 'rfaulk'}
    Dec-20 01:53:13 INFO     git_deploy.git_deploy :: Checking for lock file at stat1.wikimedia.org.
    Dec-20 01:53:15 INFO     git_deploy.git_deploy :: rfaulk has lock.
    Dec-20 01:53:15 INFO     git_deploy.git_deploy :: REVERT -> no tag specified, using: 'sartoris-sync-20131220-012354'
    Dec-20 01:53:15 INFO     git_deploy.git_deploy :: REVERT -> Attempting to revert to tag: 'sartoris-sync-20131220-012354'
    Dec-20 01:53:15 INFO     git_deploy.git_deploy :: REVERT -> Reverted to tag: 'sartoris-sync-20131220-012354', call "git deploy sync" to persist

And sync the changes:

    richrushbay-lm:test_sartoris rfaulk$ git deploy sync

    Dec-20 01:53:49 DEBUG    git-deploy is ready to run
    Dec-20 01:53:49 INFO     git_deploy.git_deploy :: Config - {'deploy.test_repo': '/tmp/test/', 'deploy.key_path': '/Users/rfaulk/.ssh/id_rsa', 'target': 'stat1.wikimedia.org', 'top_dir': '/Users/rfaulk/Projects/test_sartoris', 'user.name': 'rfaulk', 'user.email': 'rfaulk@yahoo-inc.com', 'deploy.remote_url': 'git@github.com:rfaulkner/test_sartoris.git', 'hook_dir': '.git/deploy/hooks/', 'client_path': '/Users/rfaulk/Projects/test_sartoris/', 'sync_dir': '.git/deploy/hooks//sync', 'path': '/home/rfaulk/test_sartoris/', 'repo_name': 'sartoris', 'deploy_file': '/Users/rfaulk/Projects/test_sartoris/.git/.deploy', 'user': 'rfaulk'}
    Dec-20 01:53:49 INFO     git_deploy.git_deploy :: Checking for lock file at stat1.wikimedia.org.
    Dec-20 01:53:51 INFO     git_deploy.git_deploy :: rfaulk has lock.
    Dec-20 01:53:51 INFO     git_deploy.git_deploy :: SYNC - tag local
    Dec-20 01:53:51 INFO     git_deploy.git_deploy :: SYNC - calling default sync.
    Dec-20 01:53:51 INFO     git_deploy.drivers.driver :: Calling default sync - pushing changes ...
    Dec-20 01:53:53 INFO     PUSH ->
    Dec-20 01:53:53 INFO     git_deploy.drivers.driver :: Calling default sync - pulling to target
    Dec-20 01:53:56 INFO     PULL -> Updating 779a077..38f42d3; Fast-forward
    Dec-20 01:53:56 INFO     git_deploy.git_deploy :: Checking for lock file at stat1.wikimedia.org.
    Dec-20 01:53:57 INFO     git_deploy.git_deploy :: rfaulk has lock.
    Dec-20 01:53:57 INFO     git_deploy.git_deploy :: SSH Lock destroy.

We're done, check the git history, you should see a rollback commit.  This is persisted to the remote repo and the target host:

    richrushbay-lm:test_sartoris rfaulk$ git log

    commit 38f42d3be6831c16b475786f3016bcc588499e56
    Author: rfaulk <rfaulk@yahoo-inc.com>
    Date:   Fri Dec 20 09:53:15 2013 +0000

        Rollback to sartoris-sync-20131220-012354.

    commit 779a07774dd5a7baf8dc86657cbfc491264ff970
    Author: rfaulk <rfaulk@yahoo-inc.com>
    Date:   Sun Dec 8 23:28:50 2013 -0800

        test

    commit 1ab80a78fc1e89ae6f8872282f96f5f42677b843
    Author: rfaulk <rfaulk@yahoo-inc.com>
    Date:   Sat Nov 16 18:31:26 2013 -0800

        test git-deploy 20131116_0.file.

    commit 12311863e111a266c4c1c513da45a39ed3e8cdd5
    Author: rfaulk <rfaulk@yahoo-inc.com>
    Date:   Sat Nov 2 19:42:58 2013 -0700

        Initial commit.

Finally, note the new sync tag for the rollback:

    richrushbay-lm:test_sartoris rfaulk$ git tag

    sartoris-sync-20131220-000105
    sartoris-sync-20131220-012354
    sartoris-sync-20131220-013503
    sartoris-sync-20131220-015351


CHANGES
Git-Deploy Changelog
====================


Version 0.3
-----------

    * revert (rollback) functionality
    * fix tagging model
    * dulwich dependencies for git remote push/pulls
    * driver model abstraction to allow for extensibility to other deployment models
    * split core git deploy functions into git_methods module that handles git interfacing
    * upgrades to init_gd to make tool setup more seamless
    * add paramiko support for securely talking to remotes
    * add project to pypi


Version 0.2
-----------

    * Change name of project


Version 0.1
-----------

    * Released on Dec 15th 2012, codename sartoris.
    * Initial release


git_deploy/config.py
"""
Config file for GitDeploy.
"""

__authors__ = {
    'Ryan Faulkner': 'bobs.ur.uncle@gmail.com',
    'Patrick Reilly': 'preilly@php.net',
    'Ryan Lane': 'rlane@wikimedia.org',
}
__date__ = '2013-09-08'
__license__ = 'GPL v2.0 (or later)'

from dulwich.repo import Repo
import subprocess
import sys
import logging

# Native git call
GIT_CALL = '/usr/bin/git'

# DEFAULT DEPLOY HOOKS
DEFAULT_HOOK = 'default.sync'

DEFAULT_REMOTE = 'origin'
DEFAULT_BRANCH = 'master'

DEFAULT_REMOTE_ARG_IDX = 1
DEFAULT_BRANCH_ARG_IDX = 2

# Codes emitted on exit conditions
exit_codes = {
    1: 'Operation failed.  Exiting.',
    2: 'A deployment has already been started.  Exiting.',
    3: 'Please enter valid arguments.  Exiting.',
    4: 'Missing lock file.  Exiting.',
    5: 'Could not reset.  Exiting.',
    6: 'Diff failed.  Exiting.',
    7: 'Missing tag(s).  Exiting.',
    8: 'Could not find tag.  Exiting.',
    9: 'Could not get listing from deploy target.  Exiting.',
    10: 'Please specify number of deploy tags to emit with -c.  Exiting',
    11: 'Could not find any deploys.  Exiting',
    12: 'Tagging failed. Exiting',
    13: 'Revert tag not found. Exiting',
    14: 'Commit failed, does not match HEAD. Exiting.',
    15: 'Missing git config. Aborting.',
    16: 'SSH call failed. Aborting.',
    17: 'Call to deploy hook failed. Aborting.',
    19: 'Missing system configuration item "deploy.client-path". Exiting.',
    20: 'Cannot find top level directory for the git repository. Exiting.',
    21: 'Missing system configuration item "hook-dir". Exiting.',
    22: 'Missing repo configuration item "tag-prefix". '
        'Please configure this using:'
        '\n\tgit config tag-prefix <repo>',
    23: 'Missing system configuration item "path". Exiting.',
    24: 'Missing system configuration item "user". Exiting.',
    25: 'Missing system configuration item "target". Exiting.',
    28: 'Missing system configuration item "user.name". Exiting.',
    29: 'Missing system configuration item "user.email". Exiting.',
    30: 'No deploy started. Please run: git deploy start',
    31: 'Failed to write tag on sync. Exiting.',
    32: 'Failed to write the .deploy file. Exiting.',
    33: 'git revert failed. Exiting.',
    34: 'git log failed. Exiting.',
    35: 'Revert failed, could not find tag. Exiting.',
    36: 'Not enough tags (<2) to revert on. Exiting.',
    37: 'Missing system configuration item "key-path". Exiting.',
    38: 'Missing system configuration item "test-repo-path". Exiting.',
    39: 'Call to SSH failed. Exiting.',
    40: 'Failed to run sync script. Exiting.',
    41: 'Missing system configuration item "remote-url". Exiting.',
    50: 'Failed to read the .deploy file. Exiting.',
    60: 'Invalid git deploy action. Exiting.',
}


# NullHandler was added in Python 3.1.
try:
    NullHandler = logging.NullHandler
except AttributeError:
    class NullHandler(logging.Handler):
        def emit(self, record):
            pass

# Add a do-nothing NullHandler to the module logger to prevent "No handlers
# could be found" errors. The calling code can still add other, more useful
# handlers, or otherwise configure logging.
log = logging.getLogger(__name__)
log.addHandler(NullHandler())


class GitDeployConfigError(Exception):
    """ Basic exception class for UserMetric types """
    def __init__(self, message="Git deploy error.", exit_code=1):
        Exception.__init__(self, message)
        self._exit_code = int(exit_code)

    @property
    def exit_code(self):
        return self._exit_code


def set_log(args, out, err):
    """
    Sets the logger.

    Parameters:

        args    - command line args
        out     - stdout
        err     - stderr
    """
    level = logging.WARNING - ((args.verbose - args.quiet) * 10)
    if args.silent:
        level = logging.CRITICAL + 1

    log_format = "%(asctime)s %(levelname)-8s %(message)s"
    handler = logging.StreamHandler(err)
    handler.setFormatter(logging.Formatter(fmt=log_format,
                         datefmt='%b-%d %H:%M:%S'))
    log.addHandler(handler)
    log.setLevel(level)


def configure(**kwargs):
    """ Parse configuration from git config """
    config = {}

    # Get top level directory of project
    proc = subprocess.Popen(['git', 'rev-parse', '--show-toplevel'],
                            stdout=subprocess.PIPE,
                            stderr=subprocess.PIPE)
    config['top_dir'] = proc.communicate()[0].strip()
    _repo = Repo(config['top_dir'])
    sc = _repo.get_config_stack()

    if proc.returncode != 0:
        # log.error("{0} :: {1}".format(__name__, exit_codes[exit_code]))
        raise GitDeployConfigError(message=exit_codes[20], exit_code=20)

    # Define the key names, git config names, and error codes
    config_elements = {
        'hook_dir': ('deploy', 'hook-dir', 21),
        'path': ('deploy', 'path', 23),
        'user': ('deploy', 'user', 24),
        'target': ('deploy', 'target', 25),
        'repo_name': ('deploy', 'tag-prefix', 22),
        'client_path': ('deploy', 'client-path', 19),
        'user.name': ('user', 'name', 28),
        'user.email': ('user', 'email', 29),
        'deploy.key_path': ('deploy', 'key-path', 37),
        'deploy.test_repo': ('deploy', 'test-repo-path', 38),
        'deploy.remote_url': ('deploy', 'remote-url', 41),
    }

    # Assign the values of each git config element
    for key, value in config_elements.iteritems():
        try:
            # Override with kwargs if the attribute exists
            if key in kwargs:
                config[key] = kwargs[key]
            else:
                config[key] = sc.get(value[0], value[1])
        except KeyError as e:
            log.error("{0} :: {1}".format(__name__, e.message))
            raise GitDeployConfigError(message=exit_codes[15], exit_code=15)

    config['sync_dir'] = '{0}/sync'.format(config['hook_dir'])

    config['deploy_root'] = config['client_path'] + '/.git/deploy'
    config['deploy_apps'] = config['deploy_root'] + '/apps'
    config['deploy_apps_common'] = config['deploy_root'] + '/apps/common'
    config['deploy_sync'] = config['deploy_root'] + '/sync'

    return config
